spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala

```diff
@@ -109,8 +111,180 @@ class RoiAlign[T: ClassTag] (
+      val roi_start_h = rois(offset_rois + 1) * spatial_scale - offset
+      val roi_end_w = rois(offset_rois + 2) * spatial_scale - offset
+      val roi_end_h = rois(offset_rois + 3) * spatial_scale - offset
+
+      var roi_width = roi_end_w - roi_start_w
+      var roi_height = roi_end_h - roi_start_h
+
+      if (aligned) {
+        require(roi_width >= 0 && roi_height >= 0,
+          s"ROIs in ROIAlignRotated do not have non-negative size!" +
+            s"But get ${roi_height} ${roi_width}")
+      } else {
+        roi_width = math.max(roi_width, 1.0f)
+        roi_height = math.max(roi_height, 1.0f)
+      }
+
+      val bin_size_h = roi_height / pooled_height
+      val bin_size_w = roi_width / pooled_width
+      val output_offset = n * n_stride + c * c_stride
+      val grad_output_value = gradOutputArr(output_offset + ph * h_stride + pw * w_stride)
+
+      // We use roi_bin_grid to sample the grid and mimic integral
+      val roi_bin_grid_h =
+        if (sampling_ratio > 0) sampling_ratio else math.ceil(roi_height / pooled_height).toInt
+      val roi_bin_grid_w =
+        if (sampling_ratio > 0) sampling_ratio else math.ceil(roi_width / pooled_width).toInt
+
+      // We do average (integral) pooling inside a bin
+      val count = roi_bin_grid_h * roi_bin_grid_w
+
+      for (iy <- 0 until roi_bin_grid_h) {
+        val y = roi_start_h + ph * bin_size_h + (iy + 0.5) * bin_size_h / roi_bin_grid_h
+        for (ix <- 0 until roi_bin_grid_w) {
+          val x = roi_start_w + pw * bin_size_w + (ix + 0.5) * bin_size_w / roi_bin_grid_w
+
+          val (w1, w2, w3, w4, x_low, x_high, y_low, y_high) =
+            bilinearInterpolateGradient(height, width, y.toFloat, x.toFloat)
+
+          val g1 = ev.times(grad_output_value, ev.fromType(w1 / count))
+          val g2 = ev.times(grad_output_value, ev.fromType(w2 / count))
+          val g3 = ev.times(grad_output_value, ev.fromType(w3 / count))
+          val g4 = ev.times(grad_output_value, ev.fromType(w4 / count))
+
+          if (x_low >= 0 && x_high >= 0 && y_low >= 0 && y_high >= 0) {
+            gradInputArr(gradInputOffset + y_low * width + x_low) =
+              ev.plus(gradInputArr(gradInputOffset + y_low * width + x_low), g1)
+            gradInputArr(gradInputOffset + y_low * width + x_high) =
+              ev.plus(gradInputArr(gradInputOffset + y_low * width + x_high), g2)
+            gradInputArr(gradInputOffset + y_high * width + x_low) =
+              ev.plus(gradInputArr(gradInputOffset + y_high * width + x_low), g3)
+            gradInputArr(gradInputOffset + y_high * width + x_high) =
+              ev.plus(gradInputArr(gradInputOffset + y_high * width + x_high), g4)
+          }
+        }
+      }
+    }
+  }
+
+  override def updateGradInput(input: Activity, gradOutput: Tensor[T]): Activity = {
+    require(mode == "avg", s"Only support backward for average mode, but get ${mode}")
+    val data = input.toTable[Tensor[Float]](1)
```

Comment:
Looks like the backward limited to float only?

Reply:
no, it should be classtag `T`

Reply:
Looks good to me.